### PR TITLE
make diagnostic analyzer service lazy and not require UI thread.

### PR DESCRIPTION
**Pull request by heejaechang**
### Customer scenario

A user gets better VS UI responsiveness right after solution load.

### Bugs this fixes

N/A

### Workarounds, if any

No workaround

### Risk

I don't see any. initialization that used to happen on UI thread is now happening in BG. timing is a little bit changed (since more things are now lazy), but won't be noticeable by users.

### Performance impact

a user should have less UI freeze right after solution load.

### Is this a regression from a previous update?

No

### Root cause analysis

to report errors, we need to load analyzers from the extension manager, but the manager wasn't free-thread so, we had to get that info from UI thread. depends on a number of extensions, this process could take some time. now extension manager is free-thread, we are doing the work from BG.

### How was the bug found?

Ongoing, perf investigation to improve UI delay.

### CODE REVIEW

**Comment by heejaechang** [comment]
require this from VS
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/581780?fullScreen=false

PR
https://devdiv.visualstudio.com/DevDiv/_git/VSExtensibility/pullrequest/107942?_a=overview

**Comment by heejaechang** [comment]
snippet from extension manager team.

> We did not add any new ExtensionManagerService APIs but we did make it safe to call the all of its public methods from a background thread. Modifying your code to move to background thread before invoking any ExtensionManagerService APIs should be all you need to do. You could also use the AsyncServiceProvider to get the service from a background thread as well if you aren’t currently doing so.

**Comment by heejaechang** [comment]
gave extension manager team private vsix and they tested their change with this. and it looks like we are off UI thread with these 2 changes.

**Comment by heejaechang** [comment]
@jinujoseph fixed merge conflict in case you want it for 15.8

**Comment by sharwell** [comment]
> require ExtensionManager change that got checked into 15.8

@jasonmalinowski Does this need to target **master-vs-deps** for the dependency?

**Comment by jasonmalinowski** [comment]
@sharwell Yes, it should. Good catch.

**Comment by heejaechang** [comment]
@jasonmalinowski can you take a look?

File: src/Features/Core/Portable/Diagnostics/HostAnalyzerManager.cs [hunk 1]
```diff
@@ -102,7 +96,13 @@ internal sealed partial class HostAnalyzerManager
             _hostDiagnosticAnalyzerPackageNameMap = ImmutableDictionary<DiagnosticAnalyzer, string>.Empty;
             _descriptorCache = new ConditionalWeakTable<DiagnosticAnalyzer, IReadOnlyCollection<DiagnosticDescriptor>>();
 
-            DiagnosticAnalyzerLogger.LogWorkspaceAnalyzers(hostAnalyzerReferences);
+            // DiagnosticAnalyzerLogger.LogWorkspaceAnalyzers(hostAnalyzerReferences);
```
[the remark this answers is not in the document]
Comment (heejaechang) [reply src/Features/Core/Portable/Diagnostics/HostAnalyzerManager.cs]: oops. I will uncomment it.

File: src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService.cs [hunk 2]
```diff
@@ -255,5 +255,10 @@ public bool ContainsDiagnostics(Workspace workspace, ProjectId projectId)
                 AnalyzerHelper.OnAnalyzerException_NoTelemetryLogging(ex, analyzer, diagnostic, _hostDiagnosticUpdateSource, projectId);
             };
         }
+
+        private static ImmutableArray<HostDiagnosticAnalyzerPackage> GetHostDiagnosticAnalyzerPackage(IWorkspaceDiagnosticAnalyzerProviderService diagnosticAnalyzerProviderService)
+        {
+            return (diagnosticAnalyzerProviderService != null ? diagnosticAnalyzerProviderService.GetHostDiagnosticAnalyzerPackages() : SpecializedCollections.EmptyEnumerable<HostDiagnosticAnalyzerPackage>()).ToImmutableArrayOrEmpty();
```
[the remark this answers is not in the document]
Comment (heejaechang) [reply src/Features/Core/Portable/Diagnostics/DiagnosticAnalyzerService.cs]: sure.

**Comment by jinujoseph** [comment]
test windows_debug_vs-integration_prtest

**Comment by jinujoseph** [comment]
test windows_release_vs-integration_prtest

**Comment by jinujoseph** [comment]
test windows_debug_vs-integration_prtest

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure is Jenkins test machine has latest d15.8 installed? this make extension manager to run from BG. so wondering that works with this build in jenkins or not. they checked into 15.8

**Comment by jasonmalinowski** [comment]
@heejaechang It's still 15.7. So yes, we don't expect this to pass. Not sure if we should upgrade it first; it'd be good to not drop this on the floor.

**Comment by heejaechang** [comment]
rerun tests so that it can pick up new machine which has latest d15.8 preview installed.

**Comment by heejaechang** [comment]
@jinujoseph I targeted this to 15.8.x

**Comment by jinujoseph** [comment]
@heejaechang  master-vs-deps  to dev15.8.x-vs-deps is correct

**Comment by heejaechang** [comment]
@jinujoseph can I check this in?

**Comment by jinujoseph** [comment]
in escrow will need approval.
Have tagged it , pending merging the current payload

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.preview5

**Pull request merged**